qiskit/qobj/qasm_qobj.py

```diff
@@ -0,0 +1,778 @@
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=invalid-name
+
+"""Module providing definitions of QASM Qobj classes."""
+
+import os
+
+import json
+import fastjsonschema
+
+
+path_part = 'schemas/qobj_schema.json'
+path = os.path.join(
+    os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
+    path_part)
+with open(path, 'r') as fd:
+    json_schema = json.loads(fd.read())
+validator = fastjsonschema.compile(json_schema)
+
+
+class QasmQobjInstruction:
+    """A class representing a single instruction in an QasmQobj Experiment."""
+
+    def __init__(self, name, params=None, qubits=None, register=None,
+                 memory=None, condition=None, label=None, mask=None,
+                 relation=None, val=None):
+        """Instatiate a new QasmQobjInstruction object.
+
+        Args:
+            name (str): The name of the instruction
+            params (list): The list of parameters for the gate
+            qubits (list): A list of `int`s representing the qubits the
+                instruction operates on
+            register (list): If a ``measure`` instruction this is a list
+                of `int`s containing the list of register slots in which to
+                store the measurement results (must be the same length as
+                qubits). If a ``bfunc`` instruction this is a single `int`
+                of the register slot in which to store the result.
+            memory (list): If a ``measure`` instruction this is a list
+                of `int`s containing the list of memory slots to store the
+                measurement results in (must be the same length as qubits).
+                If a ``bfunc`` instruction this is a single `int` of the
+                memory slot to store the boolean function result in.
+            condition (bool): Set to true if the gate has a conditional set
+                on it
+            label (str): An optional label assigned to the instruction
+            mask (int): For a ``bfunc`` instruction the hex value which is
+                applied as an ``AND`` to the register bits.
+            relation (str): Relational  operator  for  comparing  the  masked
+                register to the `val` kwarg. Can be either ``==`` (equals) or
+                ``!=`` (not equals).
+            val (int): Value to which to compare the masked register. In other
+                words, the output of the function is (``register AND mask)
+                relation val.
+        """
+        super(QasmQobjInstruction, self).__init__()
+        self._data = {}
+        self._data['name'] = name
```

Comment:
The purpose of `_data` is just for easy serialization, right?

Reply:
yeah, it was mostly to simplify the `to_dict()` method. The original version of this patch made each qobj object just:

class QasmQobj:

```
def __init__(self, config, header, experiments):
    self.config = config or {}
    self.header = header or {}
    self.experiments = experiments or []
```
I had to add all the class structure to make this a drop in replacement and tried to keep to that early prototype. But while I was writing this message I think it might end up being simpler to just reverse this relationship and have `to_dict()` build a dict and use attributes for the attributes.

Reply:
Done in https://github.com/Qiskit/qiskit-terra/pull/3383/commits/7ef6cffc31275c89b95c1aa82c56df07b973963b